text/0001-reactive-assignments.md

```diff
@@ -0,0 +1,1068 @@
+
+In addition, while it's possible to batch changes *within a component* (`this.set({ a, b, c })` results in a single update), changes *across* components (or those that affect stores and components together) are not batched. Now that `set` is no longer available, it is important that variable reassignments don't result in a sync update, but rather schedule an update (for the next animation frame, for example).
+
+Exceptions to this rule could be made where appropriate. For example if you're tracking the height of a list...
+
+```html
+<script>
+  import { afterUpdate } from 'svelte';
+
+  export let items = [];
+  let list;
+  let listHeight = 0;
+
+  afterUpdate(() => {
+    listHeight = list.offsetHeight;
+  });
+</script>
+
+<p>the list is {listHeight}px tall:</p>
+
+<ul ref:list>
+  {#each items as item}
+    <li>{item}</li>
+  {/each}
+</ul>
+```
+
+...then you want the change to `listHeight` to be reflected in the view as soon as all the `onupdate` callbacks have fired, *not* after an animation frame (which would result in lag).
+
+Similarly, in terms of the public component API, it might be necessary for `customElement.foo = 1` to result in a synchronous update.
+
+
+### Suspense
+
+This proposal isn't directly concerned with [implementing Suspense](https://github.com/sveltejs/svelte/issues/1736), but care should be taken to ensure that Suspense can be implemented atop this proposal without breaking changes.
+
+
+### TypeScript
+
+A major advantage of this proposal over Svelte 2 is that components become far more amenable to typechecking and other forms of static analysis. Even without TypeScript, VSCode is able to offer much richer feedback now than with Svelte 2.
+
+An eventual goal is to be able to use TypeScript in components:
+
+```html
+<script lang="typescript">
+  export let name: string;
+</script>
+
+<h1>Hello {name}!</h1>
+```
+
+Editor integrations would ideally offer autocompletion and as-you-type typechecking inside the markup. This is not my area of expertise, however, so I would welcome feedback on this proposal from people who are more familiar with the TypeScript compiler API and ecosystem.
+
+
+### Custom elements
+
+Svelte would continue to offer a custom element compile target. No real changes would be involved here, since this proposal brings 'vanilla' Svelte components in line with web components viz. property access and event handling.
+
+The `props` compiler option is redundant now that the contract is defined via `export`ed variables. The `tag` option can be set via `<svelte:meta>` as discussed above.
+
+
+### Dependency tracking
+
+In Svelte 2, 'computed properties' use compile-time dependency tracking to derive values from state, avoiding recomputation when dependencies haven't changed. These computed properties are 'push-based' rather than 'pull-based', which can result in unnecessary work:
```

Comment:
Is this section still valid with the reference implementation of #8? I think #8 is a much cleaner solution, I wonder what would be the expected behavior of function calls in html. I think the most sensible approach would be to re-run the function only when reactive arguments change.

```html
<script>
import { fn } from './functions';
</script>

<ul>
  <li>{fn()} - static, called on initial render</li>
  <li>{fn(a, b)} - dynamic, called whenever a or b change</li>
</ul>
```